test/integration/Subscription.test.js

```diff
@@ -0,0 +1,151 @@
+    let client
+    let subscription
+
+    async function setup() {
+        client = createClient()
+        client.on('error', throwError)
+        stream = await client.createStream({
+            name: uniqueId(),
+        })
+    }
+
+    async function teardown() {
+        if (subscription) {
+            await client.unsubscribe(subscription)
+            subscription = undefined
+        }
+        if (stream) {
+            await stream.delete()
+            stream = undefined
+        }
+        if (client && client.isConnected()) {
+            await client.disconnect()
+            client.off('error', throwError)
+            client = undefined
+        }
+    }
+
+    /**
+     * Returns an array which will be filled with subscription events in the order they occur.
+     * Needs to create subscription at same time in order to track message events.
+     */
+
+    function createMonitoredSubscription(opts = {}) {
+        if (!client) { throw new Error('No client') }
+        const events = []
+        subscription = client.subscribe({
+            stream: stream.id,
+            resend: RESEND_ALL,
+            ...opts,
+        }, (message) => {
+            events.push(message)
+        })
+        subscription.on('subscribed', () => events.push('subscribed'))
+        subscription.on('resending', () => events.push('resending'))
+        subscription.on('resent', () => events.push('resent'))
+        subscription.on('no_resend', () => events.push('no_resend'))
+        subscription.on('unsubscribed', () => events.push('unsubscribed'))
+        subscription.on('error', () => events.push('unsubscribed'))
+        return events
+    }
+
+    async function publishMessage() {
+        const message = {
+            message: uniqueId(),
+        }
+        await stream.publish(message)
+        return message
+    }
+
+    beforeEach(async () => {
+        await teardown()
+        await setup()
+    })
+
+    afterEach(async () => {
+        await teardown()
+    })
+
+    describe('subscribe/unsubscribe events', () => {
+        it('fires events in correct order', async (done) => {
+            const subscriptionEvents = createMonitoredSubscription()
+            subscription.on('subscribed', async () => {
+                subscription.on('unsubscribed', () => {
+                    expect(subscriptionEvents).toEqual([
+                        'subscribed',
+                        'unsubscribed',
+                    ])
+                    done()
+                })
+                await client.unsubscribe(subscription)
+            })
+
+            await client.connect()
+        })
+    })
+
+    describe('resending/no_resend events', () => {
+        it('fires events in correct order', async (done) => {
+            const subscriptionEvents = createMonitoredSubscription()
+            subscription.on('no_resend', async () => {
+                await wait(0)
+                expect(subscriptionEvents).toEqual([
```

Comment:
`resending` and `no_resend` never appear together. It's either (`resending` AND `resent`) OR `no_resend`, depending on whether messages to resend are found or not.

Reply:
Ok, that's fine, documentation in readme needs to change though.